rinja/src/filters/humansize.rs

```diff
@@ -21,24 +61,97 @@ use humansize::{DECIMAL, ISizeFormatter, ToF64};
 /// assert_eq!(tmpl.to_string(),  "Filesize: 1.23 MB.");
 /// ```
 #[inline]
-pub fn filesizeformat(b: &impl ToF64) -> Result<FilesizeFormatFilter, Infallible> {
-    Ok(FilesizeFormatFilter(b.to_f64()))
+pub fn filesizeformat<T: ToF32>(b: T) -> Result<FilesizeFormatFilter, Infallible> {
+    Ok(FilesizeFormatFilter(b.to_f32()))
 }
 
 #[derive(Debug, Clone, Copy)]
-pub struct FilesizeFormatFilter(f64);
+pub struct FilesizeFormatFilter(f32);
+
+struct NbAndDecimal(usize);
+impl NbAndDecimal {
+    fn new(nb: f32) -> Self {
+        // `nb` will never be bigger than 999_999 so we're fine with usize.
+        Self(nb as _)
+    }
+}
+impl fmt::Display for NbAndDecimal {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        let sub = self.0 % 1_000 / 10;
+        if sub != 0 {
+            if sub < 10 {
+                f.write_fmt(format_args!("{}.0{sub}", self.0 / 1_000))
+            } else {
+                f.write_fmt(format_args!("{}.{sub}", self.0 / 1_000))
+            }
+        } else {
+            f.write_fmt(format_args!("{}", self.0 / 1_000))
+        }
+    }
+}
 
 impl fmt::Display for FilesizeFormatFilter {
     fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
-        f.write_fmt(format_args!("{}", ISizeFormatter::new(self.0, &DECIMAL)))
+        if self.0 < 1_000. {
```

Comment:
I wonder if this whole `if-else-if` block could be implemented as a loop: `for (prefix, limit) in [("", 1e3), ("k", 1e6), ..] {`?

Reply:
It could, but the advantage of this approach is that we don't need to format the string of the units. :)

Reply:
I guess we could use `[" B", " kB", ..]` including the space and 'B', and have fewer write calls.

Reply:
Now I'm not sure to follow what you mean. I have in mind:

```rust
for (unit, div) in [(" KB", 1e6), ...] {
    if self.0 < div {
        f.write_fmt(format_args!("{}{unit}", NbAndDecimal::new(self.0)))
    }
}
```

That's still one extra argument formatting.

Reply:
```rust
write!(f, "{}", NbAndDecimal::new(self.0))?;
f.write_str(unit)?;
```
The extra call is there in the current implementation, too, it is just hidden in the unrolled code.

Reply:
Ok I see. Let's go with it then. Hopefully, loop will be unrolled at compile-time but it's not a hot path so should be fine.